### PR TITLE
Update XeTLA's attn implementation of Triton benchmark

### PR DESCRIPTION
**Pull request by ESI-SYD**
https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/10495671808

### CODE REVIEW

File: .github/workflows/triton-benchmarks.yml [hunk 1]
```diff
@@ -90,12 +90,25 @@ jobs:
           TRITON_INTEL_ENABLE_ADDRESS_PAYLOAD_OPT=1 \
           IGC_VISAOptions=" -TotalGRFNum 256 -enableBCR -nolocalra -printregusage -DPASTokenReduction -enableHalfLSC -abiver 2" \
           IGC_DisableLoopUnroll=1 \
-          SYCL_PROGRAM_COMPILE_OPTIONS=" -vc-codegen -vc-disable-indvars-opt -doubleGRF -Xfinalizer ' -printregusage -enableBCR -DPASTokenReduction ' " \
```
Comment (whitneywhtsang) [inline .github/workflows/triton-benchmarks.yml:93]: I thought you said this is needed for XeTLA, why is it no longer needed?

File: scripts/test-triton.sh [hunk 2]
```diff
@@ -267,17 +272,36 @@ run_benchmark_gemm() {
...
 
+run_benchmark_attention() {
+  echo "****************************************************"
+  echo "*****            Running ATTENTION             *****"
+  echo "****************************************************"
+  BENCHMARK_TEST_DIR=$TRITON_PROJ/benchmarks/triton_kernels_benchmark
+  if [ ! -d "${BENCHMARK_TEST_DIR}" ]; then
+    echo "Not found '${BENCHMARK_TEST_DIR}'." ; exit 5
+  fi
+  TRITON_INTEL_ADVANCED_PATH=1 \
```
Comment (whitneywhtsang) [inline scripts/test-triton.sh:292]: Can we change the order for the two to match the order for GEMM?

File: .github/workflows/triton-benchmarks.yml (hunk 1 above)
Comment (whitneywhtsang) [reply .github/workflows/triton-benchmarks.yml:93]: oic, as you now set it in benchmarks/xetla_kernel/CMakeLists.txt.

File: benchmarks/triton_kernels_benchmark/flash_attention_fwd_benchmark.py [hunk 3]
```diff
@@ -193,13 +187,14 @@ def forward(q, k, v, causal, sm_scale):
...
         # argument names to use as an x-axis for the plot
         x_names=['Z', 'H', 'N_CTX', 'D_HEAD'],
-        x_vals=[[4, 48, 1024, 64]],
+        x_vals=[[4, 48, 1024, 64], [32, 32, 512, 64], [16, 32, 1024, 64], [8, 32, 2048, 64], [4, 32, 4096, 64],
+                [2, 32, 8192, 64], [1, 32, 16384, 64]],
         line_arg='provider',
         # argument name whose value corresponds to a different line in the plot
         # possible values for `line_arg``
-        line_vals=['triton', 'xetla'],
+        line_vals=['xetla'],
```
Comment (whitneywhtsang) [inline benchmarks/triton_kernels_benchmark/flash_attention_fwd_benchmark.py:195]: Why are we not running Triton?

File: .github/workflows/triton-benchmarks.yml (hunk 1 above)
Comment (ESI-SYD) [reply .github/workflows/triton-benchmarks.yml:93]: yes

File: benchmarks/triton_kernels_benchmark/flash_attention_fwd_benchmark.py (hunk 3 above)
Comment (ESI-SYD) [reply benchmarks/triton_kernels_benchmark/flash_attention_fwd_benchmark.py:195]: For advanced path, I see `LLVM ERROR: TritonGPU module should contain a triton_gpu.num-warps attribute` for `Triton` attn. I think we can add back after attn features productization finished.

File: scripts/test-triton.sh (hunk 2 above)
Comment (ESI-SYD) [reply scripts/test-triton.sh:292]: Sure

File: .github/workflows/triton-benchmarks.yml [hunk 4]
```diff
@@ -90,12 +90,25 @@ jobs:
...
           python gemm_benchmark.py --reports $REPORTS
           source ../../scripts/capture-hw-details.sh
           python ../../scripts/build_report.py $REPORTS/matmul-performance.csv $REPORTS/gemm-triton-report.csv --benchmark gemm --compiler triton --param_cols "B,M,K,N" --tflops_col Triton-TFlops-max --hbm_col "Triton-GB/s-max"
           python ../../scripts/build_report.py $REPORTS/matmul-performance.csv $REPORTS/gemm-xetla-report.csv --benchmark gemm --compiler xetla --param_cols "B,M,K,N" --tflops_col XeTLA-TFlops-max --hbm_col "XeTLA-GB/s-max"
 
+      - name: Run Triton FA kernel benchmark
```
Comment (etiotto) [inline .github/workflows/triton-benchmarks.yml:98]: We need to wait to make FlashAttention run on the llvm-target branch. First we need to complete the POC "productization". Only at that point we should remove this coverage.

**Review by etiotto — CHANGES_REQUESTED** [review]

File: .github/workflows/triton-benchmarks.yml (hunk 4 above)
Comment (ESI-SYD) [reply .github/workflows/triton-benchmarks.yml:98]: OK

File: benchmarks/triton_kernels_benchmark/flash_attention_fwd_benchmark.py [hunk 5]
```diff
@@ -193,13 +187,14 @@ def forward(q, k, v, causal, sm_scale):
     benchmark_suit.Benchmark(
         # argument names to use as an x-axis for the plot
         x_names=['Z', 'H', 'N_CTX', 'D_HEAD'],
-        x_vals=[[4, 48, 1024, 64]],
+        x_vals=[[4, 48, 1024, 64], [32, 32, 512, 64], [16, 32, 1024, 64], [8, 32, 2048, 64], [4, 32, 4096, 64],
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply benchmarks/triton_kernels_benchmark/flash_attention_fwd_benchmark.py]: done.

**Review by whitneywhtsang — APPROVED** [review]

**Comment by whitneywhtsang** [comment]
Added a change to run flash attention with default path first, then we can add advanced path in a separate PR when it is ready.

**Comment by whitneywhtsang** [comment]
@ESI-SYD Please check if you are ok with https://github.com/intel/intel-xpu-backend-for-triton/pull/1877/commits/984905a97dc537ac7988f5a0c7fbdf83ea85922b.

**Comment by whitneywhtsang** [comment]
Verified that tolerance needs to be increased for advanced path in PoC branch too. https://github.com/intel/intel-xpu-backend-for-triton/pull/1877/commits/466cb15a3a24f3558db5802eb7b1fe6010ac00e0

**Review by chengjunlu — APPROVED** [review]

File: .github/workflows/triton-benchmarks.yml [hunk 6]
```diff
@@ -111,14 +110,27 @@ jobs:
...
           source ../../scripts/capture-hw-details.sh
           python ../../scripts/build_report.py $REPORTS/matmul-performance.csv $REPORTS/gemm-triton-report.csv --benchmark gemm --compiler triton --param_cols "B,M,K,N" --tflops_col Triton-TFlops --hbm_col "Triton-GB/s" --tag $TAG
           python ../../scripts/build_report.py $REPORTS/matmul-performance.csv $REPORTS/gemm-xetla-report.csv --benchmark gemm --compiler xetla --param_cols "B,M,K,N" --tflops_col XeTLA-TFlops --hbm_col "XeTLA-GB/s" --tag $TAG
 
+      - name: Run Triton FA kernel benchmark - default path
+        run: |
+          cd benchmarks/triton_kernels_benchmark
+          TRITON_INTEL_ADVANCED_PATH=0 \
+          TRITON_INTEL_ENABLE_ADDRESS_PAYLOAD_OPT=1 \
+          IGC_VISAOptions=" -TotalGRFNum 256 -enableBCR -nolocalra -printregusage -DPASTokenReduction -enableHalfLSC -abiver 2" \
```
Comment (etiotto) [inline .github/workflows/triton-benchmarks.yml:125]: We have a feature to chose the large GRF mode automatically so can you please remove the -TotalGRFNum option here ?  

We can also have 2 tests (one with extra IGC options and one without any extra options). That way we can track performance out-of-box vs with special IGC flags.

File: .github/workflows/triton-benchmarks.yml [hunk 7]
```diff
@@ -111,14 +110,27 @@ jobs:
...
+          IGC_DisableLoopUnroll=1 \
```
Comment (etiotto) [inline .github/workflows/triton-benchmarks.yml:126]: Does this option matters in the advanced path ?  Can we remove it ?
I think is important to clean up any IGC options that aren't absolutely needed to get performance.

File: scripts/test-triton.sh [hunk 8]
```diff
@@ -268,18 +273,33 @@ run_benchmark_gemm() {
   TRITON_INTEL_ENABLE_ADDRESS_PAYLOAD_OPT=1 \
   IGC_VISAOptions=" -TotalGRFNum 256 -enableBCR -nolocalra -printregusage -DPASTokenReduction -enableHalfLSC -abiver 2" \
```
Comment (etiotto) [inline scripts/test-triton.sh:274]: `-TotalGRFNum 256`  is not required for Triton, remove. 
 
 Are any of the other flags required ?

File: scripts/test-triton.sh [hunk 9]
```diff
@@ -268,18 +273,33 @@ run_benchmark_gemm() {
...
   IGC_DisableLoopUnroll=1 \
-  SYCL_PROGRAM_COMPILE_OPTIONS=" -vc-codegen -vc-disable-indvars-opt -doubleGRF -Xfinalizer ' -printregusage -enableBCR -DPASTokenReduction ' " \
   python ${BENCHMARK_TEST_DIR}/gemm_benchmark.py
 
   echo "Advanced path:"
   TRITON_INTEL_ADVANCED_PATH=1 \
   TRITON_INTEL_ENABLE_ADDRESS_PAYLOAD_OPT=1 \
   IGC_VISAOptions=" -TotalGRFNum 256 -enableBCR -nolocalra -printregusage -DPASTokenReduction -enableHalfLSC -abiver 2" \
```
Comment (etiotto) [inline scripts/test-triton.sh:281]: Remove `-TotalGRFNum 256`

File: scripts/test-triton.sh [hunk 10]
```diff
@@ -268,18 +273,33 @@ run_benchmark_gemm() {
...
+  echo "****************************************************"
+  BENCHMARK_TEST_DIR=$TRITON_PROJ/benchmarks/triton_kernels_benchmark
+  if [ ! -d "${BENCHMARK_TEST_DIR}" ]; then
+    echo "Not found '${BENCHMARK_TEST_DIR}'." ; exit 5
+  fi
+  cd $TRITON_PROJ/benchmarks; python setup.py install
+  echo "Default path:"
+  TRITON_INTEL_ADVANCED_PATH=0 \
+  TRITON_INTEL_ENABLE_ADDRESS_PAYLOAD_OPT=1 \
+  IGC_VISAOptions=" -TotalGRFNum 256 -enableBCR -nolocalra -printregusage -DPASTokenReduction -enableHalfLSC" \
```
Comment (etiotto) [inline scripts/test-triton.sh:298]: Remove ` -TotalGRFNum 256`

File: benchmarks/triton_kernels_benchmark/flash_attention_fwd_benchmark.py (hunk 3 above)
Comment (whitneywhtsang) [reply benchmarks/triton_kernels_benchmark/flash_attention_fwd_benchmark.py]: Addressed in https://github.com/intel/intel-xpu-backend-for-triton/pull/1877/commits/984905a97dc537ac7988f5a0c7fbdf83ea85922b.

File: .github/workflows/triton-benchmarks.yml (hunk 4 above)
Comment (whitneywhtsang) [reply .github/workflows/triton-benchmarks.yml]: Addressed in https://github.com/intel/intel-xpu-backend-for-triton/pull/1877/commits/984905a97dc537ac7988f5a0c7fbdf83ea85922b.

File: .github/workflows/triton-benchmarks.yml (hunk 6 above)
Comment (whitneywhtsang) [reply .github/workflows/triton-benchmarks.yml:125]: Addressed in https://github.com/intel/intel-xpu-backend-for-triton/pull/1970 and https://github.com/intel/intel-xpu-backend-for-triton/pull/1971.

File: .github/workflows/triton-benchmarks.yml (hunk 7 above)
Comment (whitneywhtsang) [reply .github/workflows/triton-benchmarks.yml:126]: Created issue https://github.com/intel/intel-xpu-backend-for-triton/issues/1972.

File: scripts/test-triton.sh (hunk 8 above)
Comment (whitneywhtsang) [reply scripts/test-triton.sh:274]: Addressed in https://github.com/intel/intel-xpu-backend-for-triton/pull/1970.
Created issue https://github.com/intel/intel-xpu-backend-for-triton/issues/1972.

File: scripts/test-triton.sh (hunk 9 above)
Comment (whitneywhtsang) [reply scripts/test-triton.sh:281]: Addressed in https://github.com/intel/intel-xpu-backend-for-triton/pull/1970.

File: scripts/test-triton.sh (hunk 10 above)
Comment (whitneywhtsang) [reply scripts/test-triton.sh:298]: Addressed in https://github.com/intel/intel-xpu-backend-for-triton/pull/1970.

**Pull request merged**